File: viz/check.mjs
// <include file="machine/comments.xml" path="//term[@id='check.mjs']"/>
// 구운 보고서가 규칙을 지켰는지 보는 검사 스크립트.
// 쓰는 것: 없음 · 쓰이는 곳: 없음
// viz/check.mjs
// 산출물 검사 규칙. 전부 기계 판정이며 사람 판단이 필요 없다.
import { readFileSync, existsSync, writeFileSync, readdirSync, rmSync } from "node:fs";
import { join, dirname } from "node:path";
import { execFileSync, spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";

const ROOT = dirname(dirname(fileURLToPath(import.meta.url)));

// <include file="machine/comments.xml" path="//term[@id='countScripts']"/>
// 산출물 HTML 의 <script> 개수를 센다. 1개까지만 허용된다.
// 쓰는 것: 없음 · 쓰이는 곳: 없음
/** <script> 는 pan/zoom 하나까지만 허용된다(산출물 불변식). */
export function countScripts(html) {
  const count = (html.match(/<script/g) || []).length;
  return { ok: count <= 1, count };
}

// <include file="machine/comments.xml" path="//term[@id='linkIntegrity']"/>
// 결정 표의 항목과 본문 절이 1:1 인지 본다.
// 쓰는 것: 없음 · 쓰이는 곳: 없음
/**
 * data.ts 의 결정 id 와 report.tsx 의 절이 1:1 인지 본다.
 * 지금은 표와 절이 어긋나도 아무도 모른다 — 이 검사가 그것을 잡는다.
 */
export function linkIntegrity(decisionIds, reportSource) {
  const sectionIds = [...reportSource.matchAll(/<Section\s+title="(D\d+)\b/g)].map((m) => m[1]);
  const missingSections = decisionIds.filter((id) => !sectionIds.includes(id));
  const orphanSections = sectionIds.filter((id) => !decisionIds.includes(id));
  return { ok: missingSections.length === 0 && orphanSections.length === 0, missingSections, orphanSections };
}


// <include file="machine/comments.xml" path="//term[@id='undefinedTerms']"/>
// 본문에 쓰였는데 용어집에 정의가 없는 식별자를 찾는다.
// 쓰는 것: C-19, calls[] · 쓰이는 곳: 없음
/**
 * 본문에 쓰인 식별자 꼴 낱말 중 용어집에 정의가 없는 것을 찾는다.
 * **경고이지 실패가 아니다** — 탐지 규칙이 오탐을 낼 수 있어 빌드를 막지 않는다.
 *
 * 잡는 꼴은 셋뿐이다. 자연어 용어(WarmUp, PageRank 같은 것)는 기계가 가릴 수 없어 빼고,
 * 저자가 용어집에 직접 넣어야 한다.
 *   - 결정 코드      C-19 · D-1 · U5 · R-13 · M4
 *   - 산출물 파일명   codegraph.json · roslyn-dump.json
 *   - 배열 필드      calls[] · edges[]
 */
export function undefinedTerms(reportSource, termIds) {
  const known = new Set(termIds);
  const found = new Set();
  const patterns = [
    /\b(?!D\d)[A-Z]{1,3}-?\d{1,3}\b/g,
    /\b[a-z][a-z0-9_-]*\.json\b/g,
    /\b[a-z][A-Za-z0-9_]*\[\]/g,
  ];
  // JSX 속성값과 import 경로는 본문이 아니다. 거칠게 걷어낸다.
  const body = reportSource
    .replace(/^import[^\n]*$/gm, "")
    .replace(/className="[^"]*"/g, "")
    // 절 제목의 D0·D1 은 결정 번호이고 링크 무결성 검사가 따로 담당한다. 통째로 뺀다.
    .replace(/<Section\s+title="[^"]*"/g, "");
  for (const re of patterns) {
    for (const m of body.matchAll(re)) {
      if (!known.has(m[0])) found.add(m[0]);
    }
  }
  return { ok: found.size === 0, missing: [...found].sort() };
}

// <include file="machine/comments.xml" path="//term[@id='versionMatch']"/>
// data.ts 의 builderVersion 이 지금 버전과 같은지 본다. 달라도 경고까지다.
// 쓰는 것: builderVersion · 쓰이는 곳: 없음
/** builderVersion 불일치는 경고이지 실패가 아니다. */
export function versionMatch(dataVersion, currentVersion) {
  return { ok: true, warn: dataVersion !== currentVersion };
}

// <include file="machine/comments.xml" path="//term[@id='check.currentBuilderVersion']"/>
// 검사 시점의 report-builder git 태그를 읽는다.
// 쓰는 것: 없음 · 쓰이는 곳: 없음
function currentBuilderVersion() {
  try {
    return execFileSync("git", ["describe", "--tags", "--abbrev=0"], {
      cwd: ROOT, stdio: ["ignore", "pipe", "ignore"],
    }).toString().trim();
  } catch {
    return "untagged";
  }
}

// 직접 실행됐을 때만 검사를 수행한다. import 시에는 순수 함수만 노출한다.
if (process.argv[1] && process.argv[1].endsWith("check.mjs")) {
  const cwd = process.cwd();
  let failed = false;

  const outFile = join(cwd, "out/report.html");
  if (!existsSync(outFile)) {
    console.error("실패 — out/report.html 이 없다. 먼저 report build 를 실행한다.");
    process.exit(1);
  }
  const html = readFileSync(outFile, "utf8");

  const s = countScripts(html);
  console.log(`${s.ok ? "통과" : "실패"} — <script> ${s.count}개 (허용 1)`);
  if (!s.ok) failed = true;

  // 타입 검사용 tsconfig 는 검사 시점에 ROOT 에 임시로 만들고 지운다.
  // 보고서마다 내용이 동일한 보일러플레이트라 대상 저장소에 남길 이유가 없다
  // (build.mjs 가 .tmp-report.mjs 를 ROOT 에 두는 것과 같은 이유).
  //
  // paths 는 타입 해결 전용이라 선언 파일을 가리킨다. .mjs 를 직접 가리키면
  // TS 가 형제 .d.mts 를 찾지 않아 TS7016 이 난다 — 런타임 해결은 build.mjs 의 alias 가 따로 한다.
  // typeRoots 를 명시하는 이유: 기본값은 tsconfig 파일 위치 기준이라 @types/node 를 놓치고 TS2688 이 난다.
  //
  // include 글로브 대신 files 에 절대경로를 열거한다. 글로브는 tsconfig 위치 기준으로
  // 해석되는데 이 파일은 ROOT 에 있고 검사 대상은 cwd 라 서로 다르다.
  const tsconfigPath = join(ROOT, ".tmp-report-tsconfig.json");
  writeFileSync(tsconfigPath, JSON.stringify({
    extends: join(ROOT, "tsconfig.json"),
    compilerOptions: {
      typeRoots: [join(ROOT, "node_modules/@types")],
      paths: {
        "report-builder": [join(ROOT, "viz/src/index.ts")],
        "report-builder/types": [join(ROOT, "viz/src/types.ts")],
        "report-builder/svg": [join(ROOT, "viz/svg.d.mts")],
      },
    },
    files: readdirSync(cwd).filter((f) => /\.tsx?$/.test(f)).map((f) => join(cwd, f)),
    include: [],
  }, null, 2) + "\n");

  let tsc;
  try {
    tsc = spawnSync("npx", ["tsc", "--noEmit", "-p", tsconfigPath], { cwd: ROOT, stdio: "pipe" });
  } finally {
    rmSync(tsconfigPath, { force: true });
  }
  console.log(`${tsc.status === 0 ? "통과" : "실패"} — tsc --noEmit`);
  if (tsc.status !== 0) {
    console.error(tsc.stdout.toString());
    failed = true;
  }

  const dataSrc = readFileSync(join(cwd, "data.ts"), "utf8");
  const reportSrc = readFileSync(join(cwd, "report.tsx"), "utf8");
  const ids = [...dataSrc.matchAll(/id:\s*"(D\d+)"/g)].map((m) => m[1]);
  const link = linkIntegrity(ids, reportSrc);
  console.log(`${link.ok ? "통과" : "실패"} — 링크 무결성 (결정 ${ids.length}건)`);
  if (!link.ok) {
    if (link.missingSections.length) console.error(`  절이 없는 결정: ${link.missingSections.join(", ")}`);
    if (link.orphanSections.length) console.error(`  결정이 없는 절: ${link.orphanSections.join(", ")}`);
    failed = true;
  }


  // 용어집 대조 — 경고만 낸다.
  // 용어는 { id: "C-19", ... } 처럼 줄 중간에 오므로 줄머리에 고정하지 않는다.
  // 결정 id(D0·D1…)는 링크 무결성 검사가 담당하므로 여기서 뺀다.
  const termIds = [...dataSrc.matchAll(/\bid:\s*"([^"]+)"/g)].map((m) => m[1]);
  const glossaryIds = termIds.filter((t) => !/^D\d+$/.test(t));
  const ut = undefinedTerms(reportSrc, glossaryIds);
  if (ut.ok) {
    console.log(`통과 — 용어집 대조 (정의 ${glossaryIds.length}개)`);
  } else {
    console.log(`경고 — 용어집에 없는 식별자 ${ut.missing.length}개`);
    console.log(`  ${ut.missing.join(", ")}`);
  }

  const dv = dataSrc.match(/builderVersion:\s*"([^"]+)"/)?.[1] ?? "?";
  const cv = currentBuilderVersion();
  const v = versionMatch(dv, cv);
  console.log(`${v.warn ? "경고" : "통과"} — builderVersion ${dv} vs ${cv}`);

  process.exit(failed ? 1 : 0);
}

File: viz/check.py
# <include file="machine/comments.xml" path="//term[@id='viz/check.py']"/>
# 구운 보고서가 규칙을 지켰는지 보는 검사 스크립트.
# 쓰는 것: 없음 · 쓰이는 곳: 없음
# viz/check.py
# 산출물 검사 규칙. 전부 기계 판정이며 사람 판단이 필요 없다.
import json
import os
import re
import subprocess
import sys
from typing import TypedDict

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


class ScriptCount(TypedDict):
    ok: bool
    count: int


class LinkResult(TypedDict):
    ok: bool
    missingSections: list[str]
    orphanSections: list[str]


class TermResult(TypedDict):
    ok: bool
    missing: list[str]


def countScripts(html: str) -> ScriptCount:
    """`<script>` 는 pan/zoom 하나까지만 허용된다(산출물 불변식)."""
    count = len(re.findall(r"<script", html))
    return {"ok": count <= 1, "count": count}


def linkIntegrity(decisionIds: list[str], reportSource: str) -> LinkResult:
    """`data.ts` 의 결정 id 와 `report.tsx` 의 절이 1:1 인지 본다."""
    sectionIds = re.findall(r'<Section\s+title="(D\d+)\b', reportSource)
    missingSections = [i for i in decisionIds if i not in sectionIds]
    orphanSections = [i for i in sectionIds if i not in decisionIds]
    return {"ok": not missingSections and not orphanSections,
            "missingSections": missingSections, "orphanSections": orphanSections}


# 본문에서 잡는 식별자 꼴 셋. `runner/term/collect.mjs` 의 findNewConcepts 와 같은 값이어야 한다.
TERM_PATTERNS = [
    re.compile(r"\b(?!D\d)[A-Z]{1,3}-?\d{1,3}\b"),
    re.compile(r"\b[a-z][a-z0-9_-]*\.json\b"),
    re.compile(r"\b[a-z][A-Za-z0-9_]*\[\]"),
]


def undefinedTerms(reportSource: str, termIds: list[str]) -> TermResult:
    """본문의 식별자 꼴 낱말 중 용어집에 정의가 없는 것을 찾는다.

    **경고이지 실패가 아니다** — 탐지 규칙이 오탐을 낼 수 있어 빌드를 막지 않는다.
    자연어 용어(WarmUp · PageRank)는 기계가 가릴 수 없어 저자가 직접 넣어야 한다.
    """
    known = set(termIds)
    found: set[str] = set()
    # JSX 속성값과 import 경로는 본문이 아니다. 거칠게 걷어낸다.
    # 절 제목의 D0·D1 은 링크 무결성 검사가 담당하므로 통째로 뺀다.
    body = re.sub(r"^import[^\n]*$", "", reportSource, flags=re.MULTILINE)
    body = re.sub(r'className="[^"]*"', "", body)
    body = re.sub(r'<Section\s+title="[^"]*"', "", body)
    for pat in TERM_PATTERNS:
        for m in pat.finditer(body):
            if m.group(0) not in known:
                found.add(m.group(0))
    return {"ok": not found, "missing": sorted(found)}


def versionMatch(dataVersion: str, currentVersion: str) -> dict[str, bool]:
    """`builderVersion` 불일치는 경고이지 실패가 아니다."""
    return {"ok": True, "warn": dataVersion != currentVersion}


def currentBuilderVersion() -> str:
    try:
        r = subprocess.run(["git", "describe", "--tags", "--abbrev=0"],
                           cwd=ROOT, capture_output=True, text=True)
        if r.returncode == 0:
            return r.stdout.strip()
    except OSError:
        pass
    return "untagged"


def main() -> int:
    cwd = os.getcwd()
    failed = False

    out_file = os.path.join(cwd, "out/report.html")
    if not os.path.exists(out_file):
        print("실패 — out/report.html 이 없다. 먼저 report build 를 실행한다.", file=sys.stderr)
        return 1
    with open(out_file, encoding="utf-8") as f:
        html = f.read()

    s = countScripts(html)
    print("%s — <script> %s개 (허용 1)" % ("통과" if s["ok"] else "실패", s["count"]))
    if not s["ok"]:
        failed = True

    # 타입 검사용 tsconfig 는 검사 시점에 ROOT 에 임시로 만들고 지운다.
    # 보고서마다 내용이 같은 보일러플레이트라 대상 저장소에 남길 이유가 없다.
    #
    # paths 는 타입 해결 전용이라 선언 파일을 가리킨다. `.mjs` 를 직접 가리키면
    # TS 가 형제 `.d.mts` 를 찾지 않아 TS7016 이 난다 — 런타임 해결은 build.py 의 alias 가 한다.
    # typeRoots 를 명시하는 이유: 기본값은 tsconfig 위치 기준이라 @types/node 를 놓치고 TS2688 이 난다.
    # include 글로브 대신 files 에 절대경로를 열거한다. 글로브는 tsconfig 위치 기준으로
    # 해석되는데 이 파일은 ROOT 에 있고 검사 대상은 cwd 라 서로 다르다.
    tsconfig_path = os.path.join(ROOT, ".tmp-report-tsconfig.json")
    with open(tsconfig_path, "w", encoding="utf-8") as f:
        json.dump({
            "extends": os.path.join(ROOT, "tsconfig.json"),
            "compilerOptions": {
                "typeRoots": [os.path.join(ROOT, "node_modules/@types")],
                "paths": {
                    "report-builder": [os.path.join(ROOT, "viz/src/index.ts")],
                    "report-builder/types": [os.path.join(ROOT, "viz/src/types.ts")],
                    "report-builder/svg": [os.path.join(ROOT, "viz/svg.d.mts")],
                },
            },
            "files": [os.path.join(cwd, n) for n in sorted(os.listdir(cwd))
                      if re.search(r"\.tsx?$", n)],
            "include": [],
        }, f, indent=2)
        f.write("\n")

    try:
        tsc = subprocess.run(["npx", "tsc", "--noEmit", "-p", tsconfig_path],
                             cwd=ROOT, capture_output=True, text=True)
    finally:
        os.remove(tsconfig_path)
    print("%s — tsc --noEmit" % ("통과" if tsc.returncode == 0 else "실패"))
    if tsc.returncode != 0:
        print(tsc.stdout, file=sys.stderr)
        failed = True

    with open(os.path.join(cwd, "data.ts"), encoding="utf-8") as f:
        data_src = f.read()
    with open(os.path.join(cwd, "report.tsx"), encoding="utf-8") as f:
        report_src = f.read()

    ids = re.findall(r'id:\s*"(D\d+)"', data_src)
    link = linkIntegrity(ids, report_src)
    print("%s — 링크 무결성 (결정 %d건)" % ("통과" if link["ok"] else "실패", len(ids)))
    if not link["ok"]:
        if link["missingSections"]:
            print("  절이 없는 결정: %s" % ", ".join(link["missingSections"]), file=sys.stderr)
        if link["orphanSections"]:
            print("  결정이 없는 절: %s" % ", ".join(link["orphanSections"]), file=sys.stderr)
        failed = True

    # 용어집 대조 — 경고만 낸다. 결정 id(D0·D1…)는 링크 무결성 검사가 담당하므로 뺀다.
    term_ids = re.findall(r'\bid:\s*"([^"]+)"', data_src)
    glossary_ids = [t for t in term_ids if not re.fullmatch(r"D\d+", t)]
    ut = undefinedTerms(report_src, glossary_ids)
    if ut["ok"]:
        print("통과 — 용어집 대조 (정의 %d개)" % len(glossary_ids))
    else:
        print("경고 — 용어집에 없는 식별자 %d개" % len(ut["missing"]))
        print("  %s" % ", ".join(ut["missing"]))

    m = re.search(r'builderVersion:\s*"([^"]+)"', data_src)
    dv = m.group(1) if m else "?"
    cv = currentBuilderVersion()
    v = versionMatch(dv, cv)
    print("%s — builderVersion %s vs %s" % ("경고" if v["warn"] else "통과", dv, cv))

    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(main())
